### DISCUSSION TITLE
Namespace field discussion for v1.1.0

### DISCUSSION THREAD

**Discussion by jsunny23** [CBIIT/ccdi-federation-api#107] [API Design Discussion]
When implementing the `namespace` fields for v1.0.0, the St. Jude Cloud team has run into an issue with our data with respect to the specification. For our namespaces, we chose to use our "data access units", which are the groupings of samples that fall under the purview of different data access committees. This is part of our file metadata, so it allows us to easily assign namespaces to files. And this concept is what most closely relates to what a user would expect as a "namespace" when working with our data.

But we have come across scenarios where an entity falls under more than one namespace. For example, a `sample` may have related `file`s that belong to different namespaces (like its restricted data access unit and feature counts for that sample in an unrestricted data access unit), and similarly a `subject` may have `sample`s that are belong in different `namespace`s.

Currently, if we have a sample that has files in two namespaces, we'd have to provide two objects in the `data` field to represent each `namespace` as the `namespace` specification only allows one `name`.

```json
"data": [
    {
        "id": {
            "namespace": {
                "organization": "stjudecloud",
                "name": "Namespace1"
            },
            "name": "Sample001"
        },
        "kind": "Participant",
        ...
    },
    {
        "id": {
            "namespace": {
                "organization": "stjudecloud",
                "name": "Namespace2"
            },
            "name": "Sample001"
        },
        "kind": "Participant",
        ...
    },
    ...
]
```

This can be confusing to a user, who sees essentially duplicate information for subjects and samples.

I have a couple of options to consider (and of course, am open to any other ideas):

### Option 1
Update the `namespace` key to `namespaces` and make it an array of objects. This would allow us to precisely define that the entity (sample or subject) has a particular `name`, and belongs to the namespaces listed within the same organization. I think this is the closest representation to our data identifiers, and my preference.

### Option 2
Add remaining namespaces to `metadata.identifiers` object. This would mean we would choose one namespace and use that under `id`, and provide the other namespace'd IDs under `metadata.identifiers`. This lets us use an existing field but will overload the usage of that field to include alternate identifiers from other organizations and identifiers within our organization

---

Another question is whether this change should be applied to all entities or not? For St. Jude Cloud data, we only *need* multiple `namespaces` for the `sample` and `subject` endpoints, but if it makes more sense to be consistent across entities, then I'm good with incorporating the change for the `file` endpoint as well.

**Comment by e-t-k** [comment]
Hi Jobin,
Thanks for raising this issue.
I agree that either **Option 1** or **Option 2** is an improvement over your current situation, which results in multiple copies of the same entity.

**Option 2** has the benefit of not requiring any interface changes and I don't see it as an "overload"; my understanding of the metadata.identifiers field has always been that it can include all alternate identifiers whether owned by the federation member or a third party.
On the other hand, this elevates one of your DAUs over the others and it may make it difficult for users to track down an entity or construct the URL for it if they're approaching it via a different DAU.

For an entity id, eg SJ031385 , is that identifier always going to refer to the same subject throughout the entire universe of St Jude? Or might SJ031385 refer to one subject in some data access units, and a different subject in other data access units?
If the identifier is unique across all of St Jude, it seems like a solution might be to make a global namespace for these identifiers eg stjudecloud/stjudecloud/SJ031385 and use that in the primary id; but I certainly don't know the details of your system.


Regarding **Option 1**, the array of namespaces - I will admit it feels very "weird". In particular, how do we construct the direct URL for an entity, currently `organization/namespace/name` ?

The best solution I can think of is that the entity wouldn't have one canonical URL but instead could be accessed by any of its namespaced identifiers - eg  https://ccdi.stjude.cloud/api/v1/subject/stjudecloud/PCGP/SJ031385  or https://ccdi.stjude.cloud/api/v1/subject/stjudecloud/ClinicalGenomics/SJ031385 would both go to the same entity. (Which is effectively how it works right now.)
And that doesn't feel too painful, actually.

Thinking about the original intent of namespaces -- to define the scope within which we can be confident that an identifier (eg SJ031385) always refers to the same entity -- **Option 1** simply defines the scope as a list of namespaces rather than a single namespace ; it feels clunky but it does have the desired effect.

And regarding which entities to affect, all "identifier" objects should definitely be treated consistently across sample/subject/file; it's just an extra hassle for clients to have to remember to check for an array or single element based on the entity type.

Ultimately my opinion relies on the answer to that question -- **are all SJ identifiers unique globally, or could SJ031385 refer to different subjects depending on the data access unit?**
If they're globally unique than **Option 2** seems a lot easier, but if you do need to distinguish on a DAU basis then **Option 1** may be necessary. (or of course some **Option 3** that's even better?)

**Comment by grugna** [comment]
I also have been of the understanding that `metadata.identifiers` would contains all other IDs whether they are internal or not, so I would be in favor of option 2.
If majority chooses to go with option 1 I am good with that one but option 2 would necessitate almost zero changes from what we have now.

**Comment by jsunny23** [comment]
Thank you both for your responses! I think either option makes sense to me, and it sounds like using `metadata.identifiers` doesn't seem like the wrong place for this use case to you both. 

Ellen - all SJ identifiers that we use for this API should be unique globally, so even if a sample belongs to two namespaces, it should be referring to the same sample. So, the risk isn't too high because a user can query either namespace and get to the same sample.
For the direct links in Option 1, I envisioned it like you mentioned. We link to our data platform, so you would essentially get the same link no matter which namespace you query from because we only use the sample name in the linked URL, and that should be the same.

I think we'll go forward with Option 2 though. It doesn't require any specification changes so we can start working on implementing it that way, and review it _if_ we encounter any issues. 

I'll leave this discussion open for a little longer in case anyone else wants to chime in, but I think I'm good with what seems to be the consensus so far!
